### PR TITLE
Opponent half obstacle

### PR DESCRIPTION
**Pull request by ichbinJosi**
## Introduced Changes

- Adds a new obstacle in the opponent half to hinder robots from walking in the opponent half when the opponent has kick-off.
- Fixes the formatting for one head
- Solves #723 

## How to Test

Check the obstacles in twix during kick-off

### CODE REVIEW

**Comment by schmidma** [comment]
This might be a dirty fix for a violating position of a supporting role. In other words, why do even try to go there in the first place? Going for a position in the opponents half should never be requested when the opponent has kick-off.

**Comment by schmidma** [comment]
I still disagree. Having a position in the opponent half in this situation is a bug. With your change this leads to path planning failing and probably the robot walking on the spot. (https://github.com/HULKs/hulk/blob/335f7a70080cd3c9585a0234b0e44921c3e17809/crates/control/src/behavior/walk_to_pose.rs#L100)

The reason penalty areas are obstacles in e.g. the penalty kick is that the robot may walk around that obstacle. In that sense I see and am okay with the plan of consistently adding every rule obstacle (and thus also the opponent half). But not in a sense of preventing a robot to plan a path to a position inside this area.

I am not sure from what motivation the original issue stems. The issue states "preventing robots from going there". While this is technically achieved, this does not fix the bug of the supporter entering the area in the first place. But reading this PR description it sounds like it does.

In short: We should open another issue to fix the bug of the supporter entering the opponent half during opponent kickoff (if that is actually the case; i could not find such a command)

**Comment by schluis** [comment]
> I still disagree. Having a position in the opponent half in this situation is a bug. With your change this leads to path planning failing and probably the robot walking on the spot. (
> 
> https://github.com/HULKs/hulk/blob/335f7a70080cd3c9585a0234b0e44921c3e17809/crates/control/src/behavior/walk_to_pose.rs#L100
> )
> 
> The reason penalty areas are obstacles in e.g. the penalty kick is that the robot may walk around that obstacle. In that sense I see and am okay with the plan of consistently adding every rule obstacle (and thus also the opponent half). But not in a sense of preventing a robot to plan a path to a position inside this area.
> 
> I am not sure from what motivation the original issue stems. The issue states "preventing robots from going there". While this is technically achieved, this does not fix the bug of the supporter entering the area in the first place. But reading this PR description it sounds like it does.
> 
> In short: We should open another issue to fix the bug of the supporter entering the opponent half during opponent kickoff (if that is actually the case; i could not find such a command)

I still disagree with your disagreement ;)

If we walk into the opponent's half during the kick-off, the robot is penalized. With this change, the robot won't be able to go there, and in the worst case, it will walk on the spot but won't be penalised. We will (hopefully) notice this and be able to fix it without loosing a robot.
You're assuming, that no one else will fuck up something before. And I don't want to assume that, I want to build a safety layer.

This kind of thinking has led the HULKs to make mistakes before. E.g. our assumption that the game controller never makes mistakes. We knew it could happen, and we chose to ignore it anyway. 
I think, we should change this mindset...

File: crates/control/src/rule_obstacle_composer.rs [hunk 1]
```diff
@@ -81,6 +80,15 @@ impl RuleObstacleComposer {
                         * context.center_circle_obstacle_increase,
                 ));
                 rule_obstacles.push(center_circle_obstacle);
+
+                let opponent_half_obstacle = RuleObstacle::Rectangle(Rectangle {
+                    min: point!(0.0, context.field_dimensions.width / 2.0),
+                    max: point!(
+                        context.field_dimensions.length / 2.0,
+                        -context.field_dimensions.width / 2.0
```
[the remark this answers is not in the document]
Comment (ichbinJosi) [reply crates/control/src/rule_obstacle_composer.rs:88]: How is max and min defined in a 2d coordinate space? We chose the x-coordinates to determine min and max. The points you chose are the same as ours just flipped.

**Comment by ichbinJosi** [comment]
> I still disagree. Having a position in the opponent half in this situation is a bug. With your change this leads to path planning failing and probably the robot walking on the spot. (
> 
> https://github.com/HULKs/hulk/blob/335f7a70080cd3c9585a0234b0e44921c3e17809/crates/control/src/behavior/walk_to_pose.rs#L100
> )
> 
> The reason penalty areas are obstacles in e.g. the penalty kick is that the robot may walk around that obstacle. In that sense I see and am okay with the plan of consistently adding every rule obstacle (and thus also the opponent half). But not in a sense of preventing a robot to plan a path to a position inside this area.
> 
> I am not sure from what motivation the original issue stems. The issue states "preventing robots from going there". While this is technically achieved, this does not fix the bug of the supporter entering the area in the first place. But reading this PR description it sounds like it does.
> 
> In short: We should open another issue to fix the bug of the supporter entering the opponent half during opponent kickoff (if that is actually the case; i could not find such a command)

We had a situation, where a robot had localization issues and ran into the opponent half. This is why we chose to add this opponent half obstacle. It is not about a robot actually planning to go there. I see that if the robot localization is wrong, the obstacle will not fix anything as it is right now. Therefore, I would suggest adding a safety margin as we did with the centre circle obstacle.

**Review by MeiiMoon — APPROVED** [review]
Lgtm
